pageserver/src/page_service.rs

```diff
@@ -316,14 +316,35 @@ impl PageServerHandler {
         Ok(())
     }
 
-    /// Helper function to wait for given lsn, or get the latest LSN if it's 0
+    /// Helper function to handle the LSN from client request.
     ///
-    /// When invalid LSN is requested, it means "don't wait, return latest version of the page"
-    /// This is necessary for bootstrap.
-    fn wait_or_get_last_lsn(timeline: &dyn Timeline, lsn: Lsn) -> Result<Lsn> {
-        if lsn == Lsn(0) {
-            Ok(timeline.get_last_record_lsn())
+    /// Wait for requested LSN to arrive, and return the LSN that should be
+    /// used to fetch the data. If 'latest' is true, the client requested for
+    /// latest page version, so we get the last record LSN and return that,
+    /// otherwise returns the request LSN.
+    ///
+    /// The requested LSN can be 0, when 'latest' is true. It means "don't
+    /// wait, fetch the latest version of the page, whatever it is". It is
+    /// used for bootstrapping.
+    fn wait_or_get_last_lsn(timeline: &dyn Timeline, lsn: Lsn, latest: bool) -> Result<Lsn> {
```

Comment:
`latest` changes semantics of lsn argument in a not extremely descriptive manner. And it is hard to grasp here what is distinction between `Lsn(0)` and `latest=true` on a first pass.

I think main problems here is that our smgr->pageserver requests are not "get page AT lsn" but rather "get page AT grater or equal lsn". "AT" scenario isn't really used now, but would be needed in future. And`latest` flag selects between that two behaviors. IMO it would be better to add enum to explicitly name that behaviors in protocol:
`get_page_at(lsn: Lsn, comparison: Equal)`
`get_page_at(lsn: Lsn, comparison: GreaterOrEqual)`
and do that both in postgres and pageserver sides.

Proper mathematical common name for '=' and ">=" is `relation`, but that is probably not a good way to name that argument in a database =). `lsn_behavior`, `lsn_operator` or just `comparison`/`comparator` are some other options.

Reply:
> `latest` changes semantics of lsn argument in a not extremely descriptive manner. And it is hard to grasp here what is distinction between `Lsn(0)` and `latest=true` on a first pass.

Fair.

> I think main problems here is that our smgr->pageserver requests are not "get page AT lsn" but rather "get page AT grater or equal lsn". "AT" scenario isn't really used now, but would be needed in future. And`latest` flag selects between that two behaviors. IMO it would be better to add enum to explicitly name that behaviors in protocol:
> `get_page_at(lsn: Lsn, comparison: Equal)`
> `get_page_at(lsn: Lsn, comparison: GreaterOrEqual)`
> and do that both in postgres and pageserver sides.

Hmm, that's not quite right either, I think. If 'latest' is true, you really want the latest version, the LSN is just a hint that there shouldn't be any version later than the given LSN. If there *is* a version later than the given lsn, something's gone wrong. In fact, it would be nice to add assertions for that. Except in the Lsn(0) case. Maybe something like this:

```
///
/// Each GetPage (and Exists and Nblocks) request includes a parameter about the
/// version of the page that is being requested. The client can request the latest
/// version of the page, or the version that's valid at a particular LSN. The primary
/// compute node will always request the latest page version, while a standby will
/// request a version at the LSN that it's currently caught up to.
///
/// In either case, if the page server hasn't received the WAL up to the requested
/// LSN yet, it will wait for it to arrive.
///
enum RequestLsn {
    /// Request latest page version. If LSN is given, it is a hint from the page
    /// server that there have been no modifications to the page after that LSN.
    /// If we haven't received WAL up to that point, will wait until we have.
    Latest(Option<Lsn>),
    /// Request page version that's valid at the given LSN.
    At(Lsn),
}
```

Reply:
> If 'latest' is true, you really want the latest version, the LSN is just a hint that there shouldn't be any version later than the given LSN. If there is a version later than the given lsn, something's gone wrong.

Yeah, I missed the fact that actually there shouldn't be any pages with bigger LSN when 'latest' is true. So 'latest' actually is a proper term, +1 on that. Also +1 on putting that lsn inside of enum instead of having a separate enum argument.

Reply:
>     /// If we haven't received WAL up to that point, will wait until we have.

I don't quite understand this comment. So if the `latestLSN` is provided, do we **have to** wait for it? 
I.e. if pageserver handles incoming WAL very slow for some reason.

Reply:
> > ```
> > /// If we haven't received WAL up to that point, will wait until we have.
> > ```
> 
> I don't quite understand this comment. So if the `latestLSN` is provided, do we **have to** wait for it?

Yes. Consider this scenario:

1. Postgres modifies page at LSN 100
2. The page is evicted from buffer cache
3. Postgres needs to read it back, so it issues a GetPage request with LSN 100

If the page server hasn't received WAL at LSN 100 yet, it must wait for it. Otherwise, it would return an obsolete page version that doesn't include the change from step 1 yet.

> I.e. if pageserver handles incoming WAL very slow for some reason.

Right.

Reply:
Thank you. I was a bit confused by the description that refers to this lsn as 'hint' .

+1 to using the `Latest(Option<Lsn>)` that you proposed in the comment above.

Reply:
So, I didn't use the `enum RequestLsn`, in the end. It felt awkward, because the enum was only used as an argument to this function. So I started working on a change where we would pass the RequestLsn down to the get_page_at_lsn() function in Timeline, so that it could check that when you request 'latest' with an LSN, there really isn't a newer page version. But that patch started to grow much larger, and that in turn was a bit awkward with the special case with 0 LSN. So I dropped that approach for now. I still think having a sanity check like that would make sense, but it's more work than I'm willing to put into this right now. I opened https://github.com/zenithdb/zenith/issues/626 to track that.